File: chapter22/bfs.go
package chapter22

import "fmt"

const (
	white = iota
	gray
	black
)

// Bfs is bfs for graph
func Bfs(graph Graph, s Vertex) {
	for i, v := range graph.table {
		if v != s {
			graph.table[i].key = white
		}
	}
	s.key = gray
	Q := make([]Vertex, len(graph.table))
	Q[0] = s
	cur := 1
	for cur != 0 {
		cur--
		u := Q[cur]
		n := u.next
		for n != nil {
			if graph.table[n.v].key == white {
				graph.table[n.v].key = gray
				Q[cur] = graph.table[n.v]
				cur++
			}
			n = n.next
		}
		fmt.Println(u.v)
		graph.table[u.v].key = black
	}
}

File: chapter22/bfs.py

from graph import initGraph


def bfs(G, s):
    G.V[s].color, G.V[s].d, G.V[s].pi = G.GRAY, 0, None
    q = [s]
    while q != []:
        u = q.pop()
        print(u)
        for v in G.Adj[u]:
            if G.V[v].color == G.WHITE:
                G.V[v].color = G.GRAY
                G.V[v].d += 1
                G.V[v].pi = u
                q.insert(0, v)
        G.V[u].color = G.BLACK


def main():
    vertex = [0, 1, 2, 3, 4, 5, 6, 7]
    edges = [(0, 1), (1, 2), (2, 3), (3, 4), (3, 5),
             (4, 5), (4, 7), (4, 6), (5, 6), (6, 7)]
    G = initGraph(vertex, edges)
    bfs(G, 3)


if __name__ == "__main__":
    main()
